commit 5c9379b14cdabc81fd5d4c916dcd93dc77c4f683
Author: Sebastian Keller <sebastian-keller@gmx.de>
Date:   Tue Sep 25 11:35:35 2012 +0200

    exa: use CLAMP_TO_EDGE for RepeatPad
    
    This fixes border rendering in some gtk3 themes.
    https://bugs.freedesktop.org/show_bug.cgi?id=55310
    
    v2 (Ben Skeggs): implement same fix for fermi/kepler too
    
    Signed-off-by: Sebastian Keller <sebastian-keller@gmx.de>
    Signed-off-by: Ben Skeggs <bskeggs@redhat.com>

diff --git a/src/nv40_exa.c b/src/nv40_exa.c
index 87ed5b2..c9e99e0 100644
--- a/src/nv40_exa.c
+++ b/src/nv40_exa.c
@@ -260,9 +260,9 @@ NV40EXAPictTexture(NVPtr pNv, PixmapPtr pPix, PicturePtr pPict, int unit)
 	if (pPict->repeat) {
 		switch(pPict->repeatType) {
 		case RepeatPad:
-			PUSH_DATA (push, NV30_3D_TEX_WRAP_S_CLAMP | 
-					 NV30_3D_TEX_WRAP_T_CLAMP |
-					 NV30_3D_TEX_WRAP_R_CLAMP);
+			PUSH_DATA (push, NV30_3D_TEX_WRAP_S_CLAMP_TO_EDGE |
+					 NV30_3D_TEX_WRAP_T_CLAMP_TO_EDGE |
+					 NV30_3D_TEX_WRAP_R_CLAMP_TO_EDGE);
 			break;
 		case RepeatReflect:
 			PUSH_DATA (push, NV30_3D_TEX_WRAP_S_MIRRORED_REPEAT |
diff --git a/src/nv50_exa.c b/src/nv50_exa.c
index d6f38eb..e70d92d 100644
--- a/src/nv50_exa.c
+++ b/src/nv50_exa.c
@@ -665,9 +665,9 @@ NV50EXAPictTexture(NVPtr pNv, PixmapPtr ppix, PicturePtr ppict, unsigned unit)
 	if (ppict->repeat) {
 		switch (ppict->repeatType) {
 		case RepeatPad:
-			PUSH_DATA (push, NV50TSC_1_0_WRAPS_CLAMP |
-				 NV50TSC_1_0_WRAPT_CLAMP |
-				 NV50TSC_1_0_WRAPR_CLAMP | 0x00024000);
+			PUSH_DATA (push, NV50TSC_1_0_WRAPS_CLAMP_TO_EDGE |
+				 NV50TSC_1_0_WRAPT_CLAMP_TO_EDGE |
+				 NV50TSC_1_0_WRAPR_CLAMP_TO_EDGE | 0x00024000);
 			break;
 		case RepeatReflect:
 			PUSH_DATA (push, NV50TSC_1_0_WRAPS_MIRROR_REPEAT |
diff --git a/src/nvc0_exa.c b/src/nvc0_exa.c
index dc8ba0b..7bd7294 100644
--- a/src/nvc0_exa.c
+++ b/src/nvc0_exa.c
@@ -659,9 +659,9 @@ NVC0EXAPictTexture(NVPtr pNv, PixmapPtr ppix, PicturePtr ppict, unsigned unit)
 		switch (ppict->repeatType) {
 		case RepeatPad:
 			PUSH_DATA (push, 0x00024000 |
-					 NV50TSC_1_0_WRAPS_CLAMP |
-					 NV50TSC_1_0_WRAPT_CLAMP |
-					 NV50TSC_1_0_WRAPR_CLAMP);
+					 NV50TSC_1_0_WRAPS_CLAMP_TO_EDGE |
+					 NV50TSC_1_0_WRAPT_CLAMP_TO_EDGE |
+					 NV50TSC_1_0_WRAPR_CLAMP_TO_EDGE);
 			break;
 		case RepeatReflect:
 			PUSH_DATA (push, 0x00024000 |
